File: packages/core/src/marketplace/registry.ts
import { MarketplaceSource, MarketplaceRegistry } from './types';

export class DefaultMarketplaceRegistry implements MarketplaceRegistry {
  private sources: MarketplaceSource[] = [];

  register(source: MarketplaceSource): void {
    this.sources.push(source);
  }

  findSource(url: string): MarketplaceSource | undefined {
    return this.sources.find((s) => s.canHandle(url));
  }

  listSources(): string[] {
    return this.sources.map((s) => s.name);
  }
}

export function createDefaultRegistry(): MarketplaceRegistry {
  const registry = new DefaultMarketplaceRegistry();
  // Adapters registered by CLI main entry
  return registry;
}

File: crates/core/src/marketplace/registry.rs
use super::types::{MarketplaceRegistry, MarketplaceSource};

pub struct DefaultMarketplaceRegistry {
    sources: Vec<Box<dyn MarketplaceSource>>,
}

impl DefaultMarketplaceRegistry {
    pub fn new() -> Self {
        DefaultMarketplaceRegistry {
            sources: Vec::new(),
        }
    }
}

impl Default for DefaultMarketplaceRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl MarketplaceRegistry for DefaultMarketplaceRegistry {
    fn register(&mut self, source: Box<dyn MarketplaceSource>) {
        self.sources.push(source);
    }

    fn find_source(&self, url: &str) -> Option<&dyn MarketplaceSource> {
        self.sources
            .iter()
            .find(|s| s.can_handle(url))
            .map(|s| s.as_ref())
    }

    fn list_sources(&self) -> Vec<&str> {
        self.sources.iter().map(|s| s.name()).collect()
    }
}

/// Create a default registry (adapters registered by CLI main entry)
pub fn create_default_registry() -> DefaultMarketplaceRegistry {
    DefaultMarketplaceRegistry::new()
}
